Respect canFuseOn{CPU,GPU} in TE fuser

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43967 Respect canFuseOn{CPU,GPU} in TE fuser**

Differential Revision: [D23469048](https://our.internmc.facebook.com/intern/diff/D23469048)

bertmaher committed:
Respect canFuseOn{CPU,GPU} in TE fuser

[ghstack-poisoned]

bertmaher committed:
Update on "Respect canFuseOn{CPU,GPU} in TE fuser"




[ghstack-poisoned]

bertmaher committed:
Update on "Respect canFuseOn{CPU,GPU} in TE fuser"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -466,6 +467,34 @@ class TensorExprFuser {
+      return true;
+    }
+    auto device = type->device();
+    if (!device) {
```
I know it's an existing logic, but I think we might want to be stricter in how we treat tensors with no device in the type, at least in the NNC fuser.

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -530,10 +559,20 @@ class TensorExprFuser {
       REQ(producer->input(1)->node()->kind() == prim::Constant);
+      auto const& listConstruct = producer->input(0)->node();
+      for (auto const& input : listConstruct->inputs()) {
+        REQ(isFusableOnDevice(input->node()));
```
Not necessarily to do in this PR, but we also need to be checking that we're pulling in values/nodes with the same device.

bertmaher replied:
I'll have to think a bit more about restricting merges to same device.  I think that problem exists even without this PR, so I'm going to go ahead and land this one while I work on same-device.

ZolotukhinM reviewed (approved):
Looks good!

bertmaher committed:
Add a negative test to show that this flag works as intended on "Respect canFuseOn{CPU,GPU} in TE fuser"




[ghstack-poisoned]

bertmaher committed:
Do not fuse if device type not set on "Respect canFuseOn{CPU,GPU} in TE fuser"




[ghstack-poisoned]

bertmaher committed:
Update on "Respect canFuseOn{CPU,GPU} in TE fuser"


Differential Revision: [D23469048](https://our.internmc.facebook.com/intern/diff/D23469048)

[ghstack-poisoned]
